src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,314 @@
+}
+
+static int query_named_solution(
+                sd_device *device,
+                const char *name,
+                const char **ret_model,
+                const char **ret_qos) {
+        _cleanup_strv_free_ char **solutions = NULL;
+        _cleanup_free_ char *upper_name = NULL, *qos_key = NULL, *model_key = NULL;
+        const char *qos = NULL, *model = NULL;
+        int r;
+
+        assert(ret_qos);
+        assert(ret_model);
+
+        /* If NULL is passed we query the default solution, which is the first one listed
+         * in the SOLUTIONS key or the one specified by the BandwidthThreshold setting.
+         */
+        if (!name) {
+                r = get_known_solutions(device, &solutions);
+                if (r < 0)
+                        return r == -ENOENT ? 0 : log_error_errno(r, "Failed to query solutions from device: %m");
+
+                r = choose_solution(solutions, &name);
+                if (r < 0)
+                        return r;
+        }
+
+        upper_name = strdup(name);
+        if (!upper_name)
+                return log_oom();
+
+        ascii_strupper(upper_name);
+        string_replace_char(upper_name, '-', '_');
+
+        qos_key = strjoin("IOCOST_QOS_", upper_name);
+        if (!qos_key)
+                return log_oom();
+
+        model_key = strjoin("IOCOST_MODEL_", upper_name);
+        if (!model_key)
+                return log_oom();
+
+        r = sd_device_get_property_value(device, qos_key, &qos);
+        if (r < 0)
+                return r == -ENOENT ? 0 : log_error_errno(r, "Failed to obtain model for iocost solution from device: %m");
+
+        r = sd_device_get_property_value(device, model_key, &model);
+        if (r < 0)
+                return r == -ENOENT ? 0 : log_error_errno(r, "Failed to obtain model for iocost solution from device: %m");
+
+        *ret_qos = qos;
+        *ret_model = model;
+
+        return 0;
+}
+
+static int apply_solution_for_path(const char *path, const char *name) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *qos = NULL, *model = NULL;
+        const char *qos_params = NULL, *model_params = NULL;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
```

Comment:
(btw, unrelated to this, I wished kernel would learn to also take a block device specification in format `<major>:<minor>:<diskseq>` format, using the new "diskseq" number that can identify a specific instance of a block device. it's really broken otherwise, since major/minor numbers are recycled aggressively, and thus you never know what you end up with)